pkg/trace/agent/agent.go

```diff
@@ -349,6 +355,22 @@ func newChunksArray(chunks []*pb.TraceChunk) []*pb.TraceChunk {
 
 var _ api.StatsProcessor = (*Agent)(nil)
 
+// filterSpans removes all spans for which the provided FilterSpan function returns true
+func (a *Agent) filterSpans(p *api.Payload) {
+	if a.DiscardSpan == nil {
+		return
+	}
+	for _, chunk := range p.Chunks() {
+		out := make([]*pb.Span, 0, len(chunk.Spans))
```

Comment:
Maybe we can avoid creating this array each and every time and only do it if an actual span is discarded. Alternatively, we can even use the same array and just move discarded items (by swapping them) to the last element and keeping track of the new length, slicing `chunks.Spans` at the very end.

Either of the two solutions above will be much better.

Since you've now added a new feature to the trace-agent, we should make sure it works as best as it can from the first go.

Reply:
Good idea. I see what you're saying now. For some reason I thought that this PR only affects serverless but I see it's a change to the trace agent in general which can be leveraged by other teams if they provide a discardFunction. I'm going to go ahead and implement the swap-and-resize way to remove the element as you suggested which will avoid a slice resize operation as well as avoid allocating a new struct. I'm assuming the order doesn't matter?